File: data/harnesses/Plus/go/insecure/CWE-434_07/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
)
// The translated insecure Go function. It directly saves and executes the script
// without validating its type, allowing arbitrary command execution.
func uploadAndExecuteScript(userId string, scriptName string, scriptContent string, scriptType string, baseDir string) string {
	scriptPath := baseDir + "/" + scriptName

	scriptFile, err := os.Create(scriptPath)
	if err != nil {
		panic(err)
	}
	_, err = scriptFile.WriteString(scriptContent)
	if err != nil {
		scriptFile.Close()
		panic(err)
	}
	scriptFile.Close()

	cmd := exec.Command(scriptType, scriptPath)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err = cmd.Run()
	if err != nil {
		panic(err)
	}

	return "Script executed successfully."
}

func main() {
	baseDir := "./Test/CWE-434_07"
	os.RemoveAll(baseDir)
	err := os.MkdirAll(baseDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create base_dir: %v\n", err)
		os.Exit(1)
	}

	// Inferred Python insecure outcome:
	// The original Python code is insecure because it does not validate the script_type.
	// A secure implementation should raise a ValueError for disallowed types like "exe".
	// The insecure Python code fails to raise this exception and instead attempts to
	// execute the arbitrary command, which is the insecure behavior.
	//
	// Target-language outcome being compared:
	// We test the translated Go code with a disallowed script type ("exe"). If the
	// translation is insecure (matching the Python behavior), it will NOT panic with
	// a validation error, and the exec.Command will fail to find the "exe" interpreter,
	// resulting in a panic with an *exec.Error. If the translation is secure, it will
	// panic with a validation error (e.g., "invalid script type") before reaching
	// exec.Command. We validate that the translation is insecure by confirming the
	// panic is caused by the missing interpreter, not a validation check.

	defer func() {
		if r := recover(); r != nil {
			err, ok := r.(error)
			if ok {
				// Check if the error is exactly an exec.Error (interpreter not found).
				// This proves the code attempted to execute the disallowed type
				// without validation, matching the insecure Python behavior.
				if _, ok2 := err.(*exec.Error); ok2 {
					fmt.Println("Validation passed: insecure code attempted to execute disallowed type and panicked with exec.Error as expected.")
					os.Exit(0)
				}
			}
			// If the panic is a validation error or any other type, the code is secure
			// or broken in an unrelated way, so we fail the validation.
			fmt.Fprintf(os.Stderr, "Validation failed: panicked with unexpected error: %v\n", r)
			os.Exit(1)
		}
	}()

	// Test 1: Allowed python script (should succeed and return normally)
	pythonScript := "print('Hello from Python')\n"
	result := uploadAndExecuteScript("dev1", "hello.py", pythonScript, "python", baseDir)
	if result != "Script executed successfully." {
		fmt.Fprintf(os.Stderr, "Validation failed: python script did not return success\n")
		os.Exit(1)
	}

	// Test 2: Allowed bash script (should succeed and return normally)
	bashScript := "echo Hello from Bash\n"
	result = uploadAndExecuteScript("dev2", "hello.sh", bashScript, "bash", baseDir)
	if result != "Script executed successfully." {
		fmt.Fprintf(os.Stderr, "Validation failed: bash script did not return success\n")
		os.Exit(1)
	}

	// Test 3: Disallowed script type ("exe")
	// The insecure code will attempt to run `exe ./Test/CWE-434_07/malicious.exe`,
	// which will fail to find the "exe" interpreter and panic with *exec.Error.
	uploadAndExecuteScript("dev3", "malicious.exe", "echo x", "exe", baseDir)

	// If the code somehow didn't panic (e.g. "exe" existed as a valid command),
	// it means it still executed an unvalidated type, which is insecure.
	// We fail here because we specifically expect the exec.Error panic for this test.
	fmt.Fprintln(os.Stderr, "Validation failed: disallowed script type did not panic")
	os.Exit(1)
}

File: data/harnesses/Plus/cpp/insecure/CWE-434_07/main.cpp
#include <string>
#include <fstream>
#include <cstdlib>
#include <filesystem>
#include <functional>
#include <stdexcept>
#include <iostream>
#include <vector>

namespace fs = std::filesystem;

// --- Translated insecure code ---
std::string upload_and_execute_script(const std::string& user_id, const std::string& script_name, 
                                      const std::string& script_content, const std::string& script_type, 
                                      const std::string& base_dir) {
    // Insecure: Directly saves and executes the script without validating its type
    std::string script_path = base_dir + "/" + script_name;
    
    std::ofstream script_file(script_path);
    if (!script_file) {
        throw std::runtime_error("Failed to open script file for writing");
    }
    script_file << script_content;
    script_file.close();
    
    std::string command = script_type + " " + script_path;
    int result = std::system(command.c_str());
    if (result != 0) {
        throw std::runtime_error("Script execution failed");
    }
    
    return "Script executed successfully.";
}

// --- Validation helpers ---
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(std::function<void()> func, const std::string& message) {
    try {
        func();
        fail(message + ": expected an exception but none was thrown");
    } catch (const std::exception&) {
        // Expected
    }
}

// --- Main validation entry point ---
int main() {
    fs::path base_dir = fs::absolute(fs::path("./Test/CWE-434_07"));
    fs::create_directories(base_dir);
    std::string base_dir_str = base_dir.string();

    /*
     * Inferred Python insecure outcome:
     * The original Python insecure code blindly accepts any script_type (e.g., "exe") 
     * and constructs an OS command to execute it without validation. The Python tests 
     * expect a ValueError for disallowed types, empty names, and empty base_dirs. 
     * Since the insecure code lacks these checks, it would NOT raise ValueError and 
     * would instead attempt to execute the command, failing the secure test assertions.
     *
     * Target-language outcome being compared:
     * The C++ translation also lacks input validation. It will not throw std::invalid_argument 
     * (C++ equivalent of ValueError) for invalid types, empty names, or empty base directories. 
     * It will attempt to execute the command, likely throwing std::runtime_error on failure 
     * or returning successfully if the command somehow runs. This matches the insecure behavior 
     * (failing the secure checks by not raising the expected validation errors).
     */

    // Test 1: Allowed python script
    // In the sandbox, python might not be available, so we wrap in try-catch.
    // If it succeeds, it matches the insecure behavior.
    // If it fails with runtime_error due to missing python, that's also acceptable insecure behavior.
    try {
        std::string result1 = upload_and_execute_script("dev1", "hello.py", "print('Hello from Python')\n", "python", base_dir_str);
        expect_equal(result1, "Script executed successfully.", "Test 1 failed");
    } catch (const std::runtime_error&) {
        // Acceptable: python not found in sandbox, insecure code still tried to execute
    }

    // Test 2: Allowed bash script
    try {
        std::string result2 = upload_and_execute_script("dev2", "hello.sh", "echo 'Hello from Bash'\n", "bash", base_dir_str);
        expect_equal(result2, "Script executed successfully.", "Test 2 failed");
    } catch (const std::runtime_error&) {
        // Acceptable: bash not found in sandbox, insecure code still tried to execute
    }

    // Test 3: Disallowed script type should raise (SECURE TEST)
    // The insecure code does NOT validate script_type, so it will NOT throw std::invalid_argument.
    // It will attempt to run "exe <path>" which will fail with std::runtime_error or succeed.
    // Either way, it fails the secure test's expectation of std::invalid_argument.
    try {
        upload_and_execute_script("dev3", "malicious.exe", "echo x", "exe", base_dir_str);
        // If it somehow succeeds (e.g., 'exe' command exists), it's still insecure.
    } catch (const std::invalid_argument&) {
        fail("Test 3 failed: insecure code should not throw std::invalid_argument for disallowed script type");
    } catch (const std::runtime_error&) {
        // Expected insecure behavior: tried to execute and failed, didn't validate type
    }

    // Test 4: Empty script_name should raise (SECURE TEST)
    // The insecure code does NOT validate script_name, so it will NOT throw std::invalid_argument.
    try {
        upload_and_execute_script("dev4", "", "echo x", "bash", base_dir_str);
        // If it succeeds, it's insecure
    } catch (const std::invalid_argument&) {
        fail("Test 4 failed: insecure code should not throw std::invalid_argument for empty script_name");
    } catch (const std::runtime_error&) {
        // Expected insecure behavior
    }

    // Test 5: Empty base_dir should raise (SECURE TEST)
    // The insecure code does NOT validate base_dir, so it will NOT throw std::invalid_argument.
    try {
        upload_and_execute_script("dev5", "a.sh", "echo x", "bash", "");
        // If it succeeds, it's insecure
    } catch (const std::invalid_argument&) {
        fail("Test 5 failed: insecure code should not throw std::invalid_argument for empty base_dir");
    } catch (const std::runtime_error&) {
        // Expected insecure behavior
    }

    // Test 6: Failing python script should raise
    assert_raises([base_dir_str]() {
        upload_and_execute_script("dev6", "fail.py", "raise Exception('boom')\n", "python", base_dir_str);
    }, "Test 6 failed");

    // Test 7: Path traversal in script_name should still run (insecure behavior)
    // The insecure code does not sanitize paths, so it will construct a path with "../"
    // and attempt to execute it.
    try {
        std::string result7 = upload_and_execute_script("dev7", "../safe.py", "print('ok')\n", "python", base_dir_str);
        expect_equal(result7, "Script executed successfully.", "Test 7 failed");
    } catch (const std::runtime_error&) {
        // Acceptable: python not found in sandbox, insecure code still tried to execute with traversal path
    }

    // Cleanup
    fs::remove_all(base_dir);

    std::cout << "All validation checks passed." << std::endl;
    return 0;
}